auth/client/iam/openid4vp.go

```diff
@@ -149,34 +154,63 @@ func (c *OpenID4VPClient) AccessToken(ctx context.Context, code string, verifier
 	// we want to make a call according to §4.1.1 of RFC6749, https://www.rfc-editor.org/rfc/rfc6749.html#section-4.1.1
 	// The URL should be listed in the verifier metadata under the "authorization_endpoint" key
 	iamClient := c.newHTTPClient()
-	metadata, err := iamClient.OAuthAuthorizationServerMetadata(ctx, verifier)
+	metadata, err := iamClient.OAuthAuthorizationServerMetadata(ctx, server)
 	if err != nil {
 		return nil, fmt.Errorf("failed to retrieve remote OAuth Authorization Server metadata: %w", err)
 	}
 	if len(metadata.AuthorizationEndpoint) == 0 {
-		return nil, fmt.Errorf("no authorization endpoint found in metadata for %s", verifier)
+		return nil, fmt.Errorf("no authorization endpoint found in metadata for %s", server)
 	}
 	endpoint, err := url.Parse(metadata.AuthorizationEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse authorization endpoint URL: %w", err)
 	}
-	// construct callback URL for wallet
-	callbackURL, err := didweb.DIDToURL(requestHolder)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create callback URL: %w", err)
+	// one default param for both signed and unsigned
+	params := map[string]interface{}{
+		oauth.ClientIDParam: client.String(),
 	}
-	callbackURL = callbackURL.JoinPath(oauth.CallbackPath)
-	redirectURL := http.AddQueryParams(*endpoint, map[string]string{
-		"client_id":     requestHolder.String(),
-		"response_type": "code",
-		"scope":         scopes,
-		"state":         clientState,
-		"redirect_uri":  callbackURL.String(),
-	})
+	// use JAR (JWT Authorization Request, RFC9101) if the verifier supports/requires it
+	if metadata.RequireSignedRequestObject {
+		// construct JWT
+		// first get a valid keyID from the vdr.KeyResolver
+		keyId, _, err := c.keyResolver.ResolveKey(client, nil, resolver.AssertionMethod)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve key for signing authorization request: %w", err)
+		}
+		// default claims for JAR
+		params[jwt.IssuerKey] = client.String()
+		params[jwt.AudienceKey] = server.String()
+		params[oauth.ClientIDParam] = client.String()
+		// added by default, can be overriden by the caller
+		params[oauth.NonceParam] = nutsCrypto.GenerateNonce()
+
+		// additional claims can be added by the caller
+		modifier(params)
+
+		token, err := c.jwtSigner.SignJWT(ctx, params, nil, keyId.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to sign authorization request: %w", err)
+		}
+		redirectURL := http.AddQueryParams(*endpoint, map[string]string{
+			oauth.ClientIDParam: client.String(),
+			oauth.RequestParam:  token,
+		})
+		return &redirectURL, nil
+	}
+	// else return an unsigned regular authorization request
+	// left here for completeness, node 2 node interaction always uses JAR since the AS metadata has it hardcoded
+
+	// additional claims can be added by the caller
+	modifier(params)
```

Comment:
modifier must be there, or can it be nil?

What I tend to do it make modifier varargs slice, then loop over it. Then the caller can either provide none, 1 or multipl.

Reply:
The instances where it is used always require a minimum of 4 options or so, never 0. A single modifier function is less code than options through var args.